### ISSUE TITLE
When adding an iOS widget extension, you can't use Flutter variables to set version

### ISSUE THREAD

**Issue by jhsware** [flutter/flutter#71262]
## Use case
When you release a Flutter app with an iOS (home screen) widget extension you can't use the Flutter variables such as $(FLUTTER_BUILD_NAME)

## Proposal
When building a Flutter project, check for iOS extensions and make the variables

$(FLUTTER_BUILD_NAME)
$(FLUTTER_BUILD_NUMBER)

etc. available to the project so archiving the iOS app doesn't require manually syncing the app version. Also, add instruction on how to set up the .plist file properly. Perhaps warning/informing during the build phase.

**Comment by jmagman** [comment]
@jhsware `FLUTTER_BUILD_NAME` and `FLUTTER_BUILD_NUMBER` are available in the `Debug.xcconfig` and `Release.xcconfig`.  You would make them the base configuration to match the Flutter app:

<img width="730" alt="Screen Shot 2020-11-30 at 5 16 55 PM" src="https://user-images.githubusercontent.com/682784/100684838-fca3ba00-332f-11eb-815a-705b9b2b53c1.png">

Then the widget extension plist versions would match the one in the Flutter app:

`CFBundleShortVersionString`=`$(FLUTTER_BUILD_NAME)`
`CFBundleVersion`=`$(FLUTTER_BUILD_NUMBER)`

![Screen Shot 2020-11-30 at 5 19 59 PM](https://user-images.githubusercontent.com/682784/100684990-4db3ae00-3330-11eb-9a9e-9650f84d8b05.png)

**Comment by jhsware** [comment]
Outstanding! I will try this.

This is one place where this documentation could be useful. And also adding a similar page for home screen widgets:
https://flutter.dev/docs/development/platform-integration/apple-watch

**Comment by jmagman** [comment]
That would be a good spot; I didn't even know that page existed.

**Comment by jmagman** [comment]
I just remembered, you may hit some issues with that https://github.com/flutter/flutter/issues/54769.  Let me know if it works.

**Comment by jhsware** [comment]
I appear to have solved this by doing the following:

- inherit from "Generated" to make sure I don't get some unknown fancy stuff from the other configs.
- manually update OTHER_LDFLAGS to remove the dependency on flutter

Now I can create an archive of both the original flutter app and the ios-extension app. 

A simple fix in Flutter would be to separate the file `Generated.xcconfig` into

- GeneralBuildVariables.xcconfig
- FlutterSpecificBuildVariables.xcconfig

and update `Debug.xcconfig` and `Release.xcconfig` to import both of those. That way an extension developer can just import `GeneralBuildVariables.xcconfig` and only get the variables needed to stay in sync with the main app.

Adding documentation here:

- https://flutter.dev/docs/development/platform-integration/apple-watch

And also adding a similar page for home screen widgets would then be sufficient IMHO.

**Comment by jmagman** [comment]
> update Debug.xcconfig and Release.xcconfig to import both of those. T

That would require a migration of these checked in files in every existing Flutter project.  We try to avoid those, inevitably someone gets into a state where everything is broken.

Let's keep this issue for documentation.  The actual code fix would be at https://github.com/flutter/flutter/issues/54769, you can follow along there.

**Comment by jhsware** [comment]
Good point! An alternative solution that would be backward compatible and probably straight forward to implement is to just generate a second file called `GeneralBuildVariables.xcconfig` that could be used by extensions. That would make the solution more solid. Even though this work around is pretty straight forward, it is a bit hackish.

**Comment by jhsware** [comment]
Fixed by https://github.com/flutter/flutter/pull/71862

**Comment by jmagman** [comment]
This still needs documentation.

**Comment by jmagman** [comment]
Sharing build configurations is now documented, but it doesn't reference the versions in the widget extension Info.plist.
https://docs.flutter.dev/platform-integration/ios/app-extensions#creating-app-extension-uis-with-flutter

**Comment by importRyan** [comment]
While the derived xcconfig absolves everyone of a .gitignore change required for a proper fix of this issue, it has a drawback of unknown future pain. All settings inside Generated.xcconfig carry forward, so one must defensively null/overwrite undesired settings (e.g., EXCLUDED_ARCHS). Since Generated.xcconfig is not checked into git, one might not notice new additions before shipping.

+1 to biting the bullet

**Comment by importRyan** [comment]
Alternative: an optional flag in pub spec to generate this additional file.
